basics/create-a-solver/create-a-solver.jmd

```diff
@@ -0,0 +1,242 @@
+---
+title: Creating a JSO compatible solver
+author: João Okimoto
+---
+
+![](../../jso-banner.png)
+
+```julia; results="hidden"
+using Pkg
+pkg"activate ."
+pkg"add NLPModels"
+pkg"add SolverTools"
+pkg"add SolverBenchmark"
+pkg"add Plots"
+pkg"add PyPlot"
+pkg"add LinearAlgebra"
+pkg"add JSOSolvers"
+pkg"add Weave"
+pkg"instantiate"
+```
+
+Julia version and JSO packages:
+```julia; echo=false
+pkgs = ["NLPModels", "SolverTools", "SolverBenchmark", "JSOSolvers"]
+
+using Pkg
+ctx=Pkg.Types.Context()
+display("text/html", "<img src=\"https://img.shields.io/badge/julia-$VERSION-3a5fcc.svg?style=flat-square&logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAA4AAAAOCAMAAAAolt3jAAAB+FBMVEUAAAA3lyU3liQ4mCY4mCY3lyU4lyY1liM3mCUskhlSpkIvkx0zlSEeigo5mSc8mio0liKPxYQ/nC5NozxQpUBHoDY3lyQ5mCc3lyY6mSg3lyVPpD9frVBgrVFZqUpEnjNgrVE3lyU8mio8mipWqEZhrVJgrVFfrE9JoTkAVAA3lyXJOjPMPjNZhCowmiNOoz1erE9grVFYqUhCnjFmk2KFYpqUV7KTWLDKOjK8CADORj7GJx3SJyVAmCtKojpOoz1DnzFVeVWVSLj///+UV7GVWbK8GBjPTEPMQTjPTUXQUkrQSEGZUycXmg+WXbKfZ7qgarqbYraSVLCUV7HLPDTKNy7QUEjUYVrVY1zTXFXPRz2UVLmha7upeMCqecGlcb6aYLWfaLrLPjXLPjXSWFDVZF3VY1zVYlvRTkSaWKqlcr6qesGqecGpd8CdZbjo2+7LPTTKOS/QUUnVYlvVY1zUXVbPSD6TV7OibLuqecGqecGmc76aYbaibLvKOC/SWlPMQjrQUEjRVEzPS0PLPDL7WROZX7WgarqibLucY7eTVrCVWLLLOzLGLCLQT0bIMynKOC7FJx3MPjV/Vc+odsCRUa+SVLCDPaWVWLKWWrLJOzPHOTLKPDPLPDPLOzLLPDOUV6+UV7CVWLKVWLKUV7GUWLGPUqv///8iGqajAAAAp3RSTlMAAAAAAAAAAAAAABAZBAAAAABOx9uVFQAAAB/Y////eQAAADv0////pgEAAAAAGtD///9uAAAAAAAAAAcOQbPLfxgNAAAAAAA5sMyGGg1Ht8p6CwAAFMf///94H9j///xiAAAw7////65K+f///5gAABjQ////gibg////bAAAAEfD3JwaAFfK2o0RAAAAAA4aBQAAABEZAwAAAAAAAAAAAAAAAAAAAIvMfRYAAAA6SURBVAjXtcexEUBAFAXAfTM/IDH6uAbUqkItyAQYR26zDeS0UxieBvPVbArjXd9GS295raa/Gmu/A7zfBRgv03cCAAAAAElFTkSuQmCC\">")
+for p in pkgs
+  uuid=ctx.env.project.deps[p]
+  v=ctx.env.manifest[uuid].version
+  c=string(hash(p) % 0x1000000, base=16)
+  display("text/html", "<img src=\"https://img.shields.io/badge/$p-$v-brightgreen?color=$c\">")
+end
+```
+
+---
+
+
+This tutorial covers the basics on how to create a solver compatible with the
+JuliaSmoothOptimizers format and how to perform basic benchmarking.
+## Newton's method implemented with the JSO API
+To illustrate the procedure for creating a solver with the JSO API, i'll be implementing
+a Line-Search Newton solver. The method consists of following
+the direction $d_{k}$ that minimizes
+$$
+m_{k}(p) = f(x_{k}) + g_{k}^{T}p + \frac{1}{2}p^{T}H_{k}p
+$$
+where $g_{k} = \nabla f(x_{k})$ and $H_{k} = \nabla f^{2}(x_{k})$. The direction $d_{k}$
+can then be written as
+$$
+d_{k} = - H_{k}^{-1}g_{k}.
+$$
+However, $d_{k}$ is only accepted if $H_{k}$ is positive definite, if not, we set
```

Comment:
In practice (and in your code), you don't check that H_k is positive definite; you check that d_k is a descent direction.

Reply:
"if not" should start a new sentence.